### DISCUSSION TITLE
Testing and managing translations

### DISCUSSION THREAD

**Discussion by Tioranat** [laravel/framework#40633] [Ideas] [answered]
I am into Laravel for some time and still keep learning. One of my recent projects is a german page for local COVID information https://www.tioranat.info

For these and other projects I had a lot of troubles checking the translations strings for the `__()` helper function.
My experience do not go along with the comments in <a href="https://github.com/laravel/framework/blob/b95b057bfa06b299309dbf65baae536104804e87/src/Illuminate/Translation/Translator.php#L132">/Illuminate/Translation/Translator.php</a>

> If the line doesn't exist, we will return back the key ... that will be quick to spot in the UI if language keys are wrong or missing.

During the development I face the following situations:
- how to build tests that all language key's are correctly set, including typos
- how to find the route where a key is used
- which keys do need to be translated still
- which packages are using their own keys

As I did not found a smart way of listening to the `__()` function my proposal would be a feature request:
Here the necessary change in `Translator.php`
```php

Illuminate/Translation/Translator.php
//
// the static storage of all translations per request
public static $log = [];

//
 
public function get( ....)
  //
  $translation = $this->makeReplacements($line ?: $key, $replace);
  self::$log[] = [$translation, $line, $key, $replace];
  return $translation;

```

To use this data you just need this code in middleware, Artisan Commands or tests:

```php
foreach(app('translator')::$log as $entry){
  [$translation, $line, $key, $replace] = $entry;
  //
  if($key === $translation){
    // translation has not worked out
  }
}


```
A test could be done like that:
```php

function testTranslationOK(){
  $response = $this->get('/');
  $response = $this->get('/legal');
  $response = $this->get('/contact');
  $response = $this->get('/blog');
  $response = $this->get('/news');

  foreach(app('translator')::$log as $entry){
    [$translation, $line, $key, $replace] = $entry;
    $this->assertNotEqual($translation, $key);
  }
}

// with some more lines the static $log could be wrapped in a method
app('translator')::getLog();

```
What do you think, is this a feature Laravel developers would love?
Are there better solutions available?

**Comment by barryvdh** [comment, marked as answer]
Replying here because you pinged me with an email about Debugbar, but have you seen https://github.com/barryvdh/laravel-translation-manager#detect-missing-translations ? That replaces the Translator and logs them; https://github.com/barryvdh/laravel-translation-manager/blob/master/src/Translator.php

**Reply by Tioranat** [reply]
Thanks for this link, I was looking for that
